### PR TITLE
Return subscription status to UI when unapproved user is trying to login

### PR DESCRIPTION
**Pull request by alexeykazakov**
/api/login redirects unapproved users to `https://openshift.io/not-authorized.html?status=${status}`
Where `${status}` can be:
- `awaiting_approval`
- `provision_queued`
- `signup_needed`

Part of https://github.com/fabric8-services/fabric8-auth/issues/419

### CODE REVIEW

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-422-1`

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-422-2`

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-422-3`

File: configuration/configuration.go [hunk 1]
```diff
@@ -562,7 +569,6 @@ func (c *ConfigurationData) setConfigDefaults() {
 	c.v.SetDefault(varGitHubClientSecret, defaultGitHubClientSecret)
 	c.v.SetDefault(varGitHubClientDefaultScopes, "admin:repo_hook read:org public_repo read:user")
 	c.v.SetDefault(varOSOClientApiUrl, "https://api.starter-us-east-2.openshift.com")
-	c.v.SetDefault(varTLSInsecureSkipVerify, false) // Do not set to true in production! True can be used only for testing.
```
Comment (sbose78) [inline configuration/configuration.go]: Why are we removing this? Might be handy for setting up on minishift.

Comment (alexeykazakov) [reply configuration/configuration.go]: Just feel uncomfortable having it around... and we don’t use it atm.

File: login/oso_reg_app.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			"err":         err,
+			"reg_app_url": regAppURL,
+			"body":        bodyString,
+		}, "unable to unmarshal json with subscription status")
+		return "", autherrors.NewInternalError(ctx, err)
+	}
+
+	for _, subscription := range sbs.Subscriptions {
+		if config.GetOSOClusterByURL(subscription.Plan.Service.APIURL) != nil {
+			return subscription.Status, nil
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply login/oso_reg_app.go]: Yes. We return whatever status is in OSO but we don't expect anything but awaiting_approval or provision_queued here.

File: login/service.go [hunk 3]
```diff
@@ -303,6 +312,16 @@ func (keycloak *KeycloakOAuthProvider) CreateOrUpdateIdentityAndUser(ctx context
 
 			userNotApprovedRedirectURL := config.GetNotApprovedRedirect()
 			if userNotApprovedRedirectURL != "" {
+				status, err := keycloak.osoSubscriptionManager.LoadOSOSubscriptionStatus(ctx, config, *keycloakToken)
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply login/service.go]: Correct.

File: login/oso_reg_app.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+type osoRegistrationApp struct {
+	httpClient rest.HttpClient
+}
+
+// NewOSORegistrationApp constructs a new OSOSubscriptionManager with default HTTP Client
+func NewOSORegistrationApp() OSOSubscriptionManager {
+	return &osoRegistrationApp{httpClient: http.DefaultClient}
+}
+
+func NewOSORegistrationAppWithClient(client rest.HttpClient) OSOSubscriptionManager {
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply login/oso_reg_app.go]: Both options are fine (one constructor with nullable param or two constructors). It's just my personal preference to keep them separate. Both are covered by tests, so, we should be fine.

**Comment by alexeykazakov** [comment]
> Have you tested the workflow ?

Yes, I have.

**Comment by sbose78** [comment]
Observation:

```
ERRO[2018-04-09 12:54:45] unable to load OSO subscription status        file="/home/shbose/fabric8/auth/auth2/src/github.com/fabric8-services/fabric8-auth/login/oso_reg_app.go" func="github.com/fabric8-services/fabric8-auth/login.(*osoRegistrationApp).LoadOSOSubscriptionStatus" line=100 pid=23006 pkg=login reg_app_url="https://manage.openshift.com/api/accounts/shXXXXXXXXXXXXXX0/subscriptions?authorization_username=sXXXXXXXXXXXe" req_headers=map[Connection:[keep-alive] User-Agent:[Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36] Upgrade-Insecure-Requests:[1] Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8] Accept-Encoding:[gzip, deflate, br] Accept-Language:[en-US,en;q=0.9]] req_id=UReV8sxTvq-2 
 response_body="{\"error\":\"Invalid Authorization Token\"}" response_status="401 Unauthorized"
```

Looks like the token didn't work in this case, but I got redirected to `https://openshift.io/not-authorized.html?status=`

**Comment by sbose78** [comment]
Run 

```
AUTH_DEVELOPER_MODE_ENABLED=true AUTH_OSO_REGAPP_ADMIN_USERNAME=XXXX AUTH_OSO_REGAPP_ADMIN_TOKEN=XXXXXXXXXXXXXXXXX
AUTH_OSO_REGAPP_SERVICEURL=XXXXXXXXXXXXXX
AUTH_NOTAPPROVED_REDIRECT=https://openshift.io/not-authorized.html ./bin/auth
```
1. Create RHD account.
2. Login to `localhost:8089/api/login?redirect=......` redirects to `https://openshift.io/not-authorized.html?status=signup_needed`
3. Sign up for `us-starter-east-2a` using the reg-app url.
4. Login to `localhost:8089/api/login?redirect=......` redirects to `https://openshift.io/not-authorized.html?status=awaiting_approval`
5. Approve account on reg-app
6. Login to `localhost:8089/api/login?redirect=......` redirects to `https://openshift.io/not-authorized.html?status=` because user is not approved in keycloak
7. Approve on keycloak
8. Login to `localhost:8089/api/login?redirect=......` is successful.

**Pull request merged**